File: ios/NfcHandler.swift
import Flutter
import CoreNFC

@available(iOS 13.0, *)
class NfcHandler: NSObject, NFCNDEFReaderSessionDelegate {
    private var flutterResult: FlutterResult?
    private var methodChannel: FlutterMethodChannel
    private var session: NFCNDEFReaderSession?
    
    init(methodChannel: FlutterMethodChannel) {
        self.methodChannel = methodChannel
        super.init()
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isAvailable":
            result(NFCNDEFReaderSession.readingAvailable)
        case "startSession":
            startSession(result: result)
        case "stopSession":
            stopSession(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func startSession(result: @escaping FlutterResult) {
        guard NFCNDEFReaderSession.readingAvailable else {
            result(FlutterError(code: "UNAVAILABLE", message: "NFC not available", details: nil))
            return
        }
        
        session = NFCNDEFReaderSession(delegate: self, queue: nil, invalidateAfterFirstRead: false)
        session?.alertMessage = "Hold your iPhone near the NFC tag"
        session?.begin()
        result(nil)
    }
    
    private func stopSession(result: @escaping FlutterResult) {
        session?.invalidate()
        session = nil
        result(nil)
    }
    
    // NFCNDEF delegate methods
    func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFs messages: [NFCNDEFMessage]) {
        guard let message = messages.first, let record = message.records.first else { return }
        
        if let payload = String(data: record.payload, encoding: .utf8) {
            methodChannel.invokeMethod("onTagRead", arguments: payload)
        }
    }
    
    func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
        // Handle session invalidation
    }
}

File: android/app/src/main/kotlin/com/grapph/allcoderelay/NfcHandler.kt
package com.grapph.allcoderelay

import android.app.Activity
import android.nfc.NfcAdapter
import android.nfc.Tag
import android.nfc.tech.Ndef
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class NfcHandler(private val activity: Activity, private val channel: MethodChannel) : NfcAdapter.ReaderCallback {
    private var nfcAdapter: NfcAdapter? = null
    
    init {
        nfcAdapter = NfcAdapter.getDefaultAdapter(activity)
    }
    
    fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isAvailable" -> result.success(nfcAdapter != null)
            "startSession" -> startSession(result)
            "stopSession" -> stopSession(result)
            else -> result.notImplemented()
        }
    }
    
    private fun startSession(result: MethodChannel.Result) {
        if (nfcAdapter == null) {
            result.error("UNAVAILABLE", "NFC not available", null)
            return
        }
        
        try {
            nfcAdapter?.enableReaderMode(activity, this, 
                NfcAdapter.FLAG_READER_NFC_A or 
                NfcAdapter.FLAG_READER_NFC_B or 
                NfcAdapter.FLAG_READER_NFC_F or 
                NfcAdapter.FLAG_READER_NFC_V, null)
            result.success(null)
        } catch (e: Exception) {
            result.error("ERROR", e.message, null)
        }
    }
    
    private fun stopSession(result: MethodChannel.Result) {
        try {
            nfcAdapter?.disableReaderMode(activity)
            result.success(null)
        } catch (e: Exception) {
            result.error("ERROR", e.message, null)
        }
    }
    
    override fun onTagDiscovered(tag: Tag) {
        val ndef = Ndef.get(tag)
        ndef?.let {
            val ndefMessage = it.cachedNdefMessage
            ndefMessage?.let { message ->
                if (message.records.isNotEmpty()) {
                    val record = message.records[0]
                    val payload = String(record.payload)
                    activity.runOnUiThread {
                        channel.invokeMethod("onTagRead", payload)
                    }
                }
            }
        }
    }
}